### PR TITLE
Experiment with a processor that finds all catalog-info.yaml files in a repository

### PR DESCRIPTION
**Pull request by Fox32**
We have bigger mono repositories with a lot of components and teams are wondering if they really have to register every `catalog-info.yaml` file in Backstage. We thought a little bit about it and build a small processor that automatically registers all `catalog-info.yaml` files in a repository. This is not only done once, but on every refresh which means new catalog-info.yaml files are automatically detected and registered.

As this solution has some drawbacks, I would like to get some feedback before finishing this PR. There are also some implementation questions I have.

I think the biggest drawback is that we have to download the full repository archive every time just to find the catalog-info.yaml, which can be slow. We don't even read the contents but are only interested in the directory tree! I tried this with one of our bigger ones and it takes more than a minute.

I left some additional comments in the code, please have a look.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by freben** [comment]
I'm off this week so not reviewing fully now, but first thought is that one would like to check if the main branch head has moved since last time. If it hasn't, there would be no need to look for changes.

File: plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    const gitUrl = parseGitUri(target);
+    const baseUrl = stringify(gitUrl);
+    // TODO: The GitHubReader doesn't know the ref yet if it's missing.
+    //  We could extend it by using the github API to retrieve the default
+    //  branch. On the other hand I'm not sure where the "..archive/${ref}.tar.gz"
+    //  url is comming from. The official endpoint says that REF is optional:
+    //  https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#download-a-repository-archive-tar
+    const ref = 'master';
+
+    return `${baseUrl}/blob/${ref}/${file}`;
```
Comment (adamdmharvey) [inline plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts:110]: Elsewhere in the code base I think there are other references to it, but I believe this format only works on GitHub and possibly GitLab, but does NOT work on BitBucket?  Wonder if we need a utility function somehow to align finding a file in a branch by Git provider?  (even if a future iteration)

**Review by adamdmharvey — COMMENTED** [review]
Like the idea!  Some feedback, welcome to consider for iterations.

File: plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    if (location.type !== 'url') {
+      return false;
+    }
+
+    const pattern = this.options.pattern ?? defaultPattern;
+    const { filepath } = parseGitUri(location.target);
+
+    // TODO: Maybe find a better way to detect whether discovery is necessary
+    //  (e.g. by trying to download it, or by reading the tree first?)
+    if (pattern.test(filepath ?? '')) {
```
Comment (Fox32) [inline plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts:51]: Here we could also use `parseGitUri(target).filepathtype ? 'repo' : 'file'`

File: plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts (hunk 1 above)
Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts:110]: Yeah, this is provide specific, see the TODO comments

File: plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  //  that all existing readers have to fill.
+  private async generateUrl(target: string, file: string): Promise<string> {
+    const gitUrl = parseGitUri(target);
+    const baseUrl = stringify(gitUrl);
+    // TODO: The GitHubReader doesn't know the ref yet if it's missing.
+    //  We could extend it by using the github API to retrieve the default
+    //  branch. On the other hand I'm not sure where the "..archive/${ref}.tar.gz"
+    //  url is comming from. The official endpoint says that REF is optional:
+    //  https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#download-a-repository-archive-tar
+    const ref = 'master';
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts:108]: One thing I've been pondering lately is whether locations will eventually need to be able to be made parmeterized. Essentially being able to have the parameter `branchName?: string` in your location that points to a repo.

[the remark this answers is not in the document]
Comment (andrewthauer) [reply plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts:108]: I might be missing something but can't the `ref` just be extracted via `parseGitUrl`?

[the remark this answers is not in the document]
Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts:108]: No, it should be that easy. But it's something we can fix on e we discussed the more genal topics 😉 we should also add tests before merging and clean everything up.

**Comment by nirga** [comment]
Looks super interesting and kinda similar to what I proposed in #3802. Some thoughts:

1. Maybe we can extend this to scan multiple repos? For example, we have ~900 repos, each with a few yamls we want to add to Backstage so it'll be useful if this can be used for that.
2. I'd opt to use Github's GQL API for that. I've done that internally as a separate crawler that finds all yaml files and then output them into a `app-config.yaml` and it's clean and neat IMHO.

**Comment by Fox32** [comment]
> I'd opt to use Github's GQL API for that. 

GraphQL would be interesting. The idea was to use `readTree` as it works across different providers. The GraphQL solution would be GitHub specific. But we could add another feature like `listTree` to the UrlReader and only allows to search for files (e.g. specific folders, file types), without content access. This is either based on `readTree` or a specialized version for GitHub that uses GraphQL.

> Maybe we can extend this to scan multiple repos

Scanning repos would also be provider specific. However it could be in a separate processor that generates the locations consumed by this processor.

File: plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts (hunk 2 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts:51]: As seen in https://github.com/backstage/backstage/pull/4014 - the filepathtype thing is not a particularly good check it seems, for some providers at least

**Comment by freben** [comment]
Before I look in depth at the code, what would you think about giving good old `UrlReaderProcessor` the ability to handle globs instead? So you'd register `https://github.com/backstage/backstage/**/catalog-info.yaml`, or even `https://github.com/backstage/backstage/**` and it being smart enough to figure out the rest.

Yes that would probably benefit heavily from a `listTree` or similar.

**Comment by Fox32** [comment]
Yeah that would be an option, too. First I thought that it would be the only option to include it there, but I managed to pull it out. Globs would be handy to provide everyone more freedom on the `catalog-info.yaml` choice.

> Before I look in depth at the code

I think a deep look isn't necessary at this stage, I'm more looking for the high level discussions that we are having right now. The input is already quite valuable and I have to iterate first. I move this to be a draft and have a look at something like `listTree` first.

**Comment by nirga** [comment]
Hey @Fox32 are you working on something in this area? I'd love to tackle what @freben suggested about extending `UrlReaderProcessor` which will allow me to later add an ability for automatically scanning all of the org's repos which I'm desperately missing at this point :)

**Comment by Fox32** [comment]
I already startet to implement some of the suggestions, but won't come back to this topic till later this week.

**Comment by nirga** [comment]
Thanks @Fox32! I want to work towards #3802 so I'll work on the feeding part of this - so, extending `GithubOrgReaderProcessor` to support outputting location entities with all repositories in a Github org (also posted about it on Discord).

**Comment by Fox32** [comment]
Closing this in favor of #4397

**Pull request closed without merging**